grafana/grafana
# Explore: deprecate compact URL encoding & dafault to extended

**What this PR does / why we need it**:
Removes compact query string encoding in explore and defaults to the extended encoding.

Currently, Explore encodes pane states in arrays in the `left` and `right` parameters.
Items in this array have different meanings based on their position:
```
[timeFrom, timeTo, datasource, ...queries[, extraParameters]]
```

This is problematic when injecting `extraParameters` as, in order to check whether an object is a query or an "extra parameter", we need to check for the presence of specific keys and it's not very scalable. 
For example in https://github.com/grafana/grafana/pull/44017 we introduced a `panelsState` object in the query string, and because of the logic mentioned above, to avoid potential conflicts with queries that may have the same key in their encoded representation, we encode this parameter as `__panelsState`. see https://github.com/grafana/grafana/pull/44017/files#diff-ccef2c5db041700ad4f35f71fdb30d59d9d23f6d02c4877704cf7687eb1a5b8aR253 and https://github.com/grafana/grafana/pull/44017/files#diff-c1f9f4853235e6df01609aa6f51331a77beaad49fb3ba578e5a523390bd927ecR209-R214.
While the solution is ok, other datasource plugins may have queries with the same key, making said datasources unusable I explore.

This PR deprecates the compact encoding by always using the extended version (a normal JSON representation of the state) when generating Explore URLs so that properties are explicitly keyed, for example:
```
/explore?left={"datasource":"gdev-testdata","queries":[{"refId":"A"}],"range":{"from":"now-1h","to":"now"}}
```

the compact encoding is still supported when accessing explore via a manually inputted URL in order to maintain compatibility with URLs generated prior to this change, but it won't be used to generate new URLs.

**How to test**:
Run a previous grafana version that generates compact explore URLs:
```
devenv sources=grafana grafana_version=8.3.4
```
and play a bit with Explore.
Then copy URLs generated from the container and paste them in the version from this branch and it should open the same Explore pages.

(opening an URL with a datasource not configured in one of the instance should fallback to the current logic of selecting a datasource in explore (`last used datasource || default datasource` ))

When pasting URLs in the new version, the resulting URL in the browser should be transformed in the non-compact version
(ie. using `/explore?left=["now-1h","now","Tempo",{"refId":"A"}]` should result in `/explore?left={"datasource":"gdev-testdata","queries":[{"refId":"A"}],"range":{"from":"now-1h","to":"now"}}`

Also, URLs should work both ways, so also URLs generated from this branch should work in 8.3.4

**Other notes**:
- In query history only queries are saved, every other parameter is ignored. we may want to consider changing this behavior in the future
- shortened links save the whole URL so they should work with other parameters as well. to test them you can add `,"panelsState":{"trace":{"spanId":"1"}}` inside the `left` or `right` query string parameters (this was introduced in https://github.com/grafana/grafana/pull/44017).
